packages/core/src/new-api/internal/reconciliation/execution-state-resolver.ts

```diff
@@ -14,12 +15,59 @@ import {
   StaticCallExecutionState,
 } from "../types/execution-state";
 import { isAddress } from "../utils";
+import { replaceWithinArg } from "../utils/replace-within-arg";
 
 import { ReconciliationContext } from "./types";
 import { resolveModuleParameter, safeToString } from "./utils";
 
 // TODO: consider merging this into the execution state map
 export class ExecutionStateResolver {
+  public static resolveArgsFromExectuionState(
+    constructorArgs: ArgumentType[],
+    context: ReconciliationContext
+  ): ArgumentType[] {
+    return replaceWithinArg<ArgumentType>(constructorArgs, {
+      bigint: (bi) => bi.toString(),
+      future: (f) => {
+        if (!isContractFuture(f)) {
+          throw new IgnitionError(
+            `Cannot replace future in args, for non-deployable futures ${f.id}`
+          );
+        }
+
+        return ExecutionStateResolver.resolveContractToAddress(f, context);
```

Comment:
I believe this can try to resolve futures that were not successfully executed. I couldn't find a check for that.

Reply:
The `resolveContractToAddress` function has an _assertion_ that a contract address is returned, so we will exception for a missing address with a relevant error:

https://github.com/NomicFoundation/ignition/blob/a65a88667d763b1f7f4e08fe416423861584ec68/packages/core/src/new-api/internal/reconciliation/execution-state-resolver.ts#L84

But my use of an assertion implies a stronger check earlier on - that you can't have a dependency on an unstarted future from an already started future. We have this check, but I am now worried about the ordering of the checks.

Let me add a test for this to ensure we check dependencies before we need to do this resolution.

Reply:
Resolution of arguments happens after dependency checks:

https://github.com/NomicFoundation/ignition/blob/82ecb97f1db8cac47f545d8d8ea61cf372775193/packages/core/src/new-api/internal/reconciliation/reconciler.ts#L32

I have assumed we can't get a request of resolution of an address on a future that has started but not completed.

If `A <- B`, where A is a call and b depends on its result as a constructor arg, B isn't allowed to start until A is complete. B cannot have execution state, hence we never reconcile it. You could change a dep to create this situation, but our dependency check will ensure we don't have both A and B having started.